apps/client/src/features/viewers/lower-thirds/LowerThird.tsx

```diff
@@ -0,0 +1,243 @@
+import { useEffect, useMemo, useState } from 'react';
+import { useSearchParams } from 'react-router-dom';
+import { Message, OntimeEvent, ViewSettings } from 'ontime-types';
+
+import { overrideStylesURL } from '../../../common/api/apiConstants';
+import NavigationMenu from '../../../common/components/navigation-menu/NavigationMenu';
+import { LOWER_THIRD_OPTIONS } from '../../../common/components/view-params-editor/constants';
+import ViewParamsEditor from '../../../common/components/view-params-editor/ViewParamsEditor';
+import { useRuntimeStylesheet } from '../../../common/hooks/useRuntimeStylesheet';
+
+import './LowerThird.scss';
+
+enum SrcKeys {
+  Title = 'title',
+  Subtitle = 'subtitle',
+  Presenter = 'presenter',
+  LowerMsg = 'lowerMsg',
+}
+
+enum TriggerType {
+  Event = 'event',
+  Manual = 'manual',
+}
+
+type LowerOptions = {
+  width: number;
+  trigger: TriggerType;
+  topSrc: SrcKeys;
+  bottomSrc: SrcKeys;
+  topColour: string;
+  bottomColour: string;
+  topBg: string;
+  bottomBg: string;
+  topSize: string;
+  bottomSize: string;
+  transition: number;
+  delay: number;
+  key: string;
+  lineColour: string;
+  lineHeight: string;
+};
+
+interface LowerProps {
+  eventNow: OntimeEvent | null;
+  viewSettings: ViewSettings;
+  lower: Message;
+}
+
+const defaultOptions: Readonly<LowerOptions> = {
+  width: 45,
+  trigger: TriggerType.Event,
+  topSrc: SrcKeys.Title,
+  bottomSrc: SrcKeys.Subtitle,
+  topColour: '000000ff',
+  bottomColour: '000000ff',
+  topBg: '00000000',
+  bottomBg: '00000000',
+  topSize: '65px',
+  bottomSize: '40px',
+  transition: 3,
+  delay: 3,
+  key: 'ffffffff',
+  lineColour: 'ff0000ff',
+  lineHeight: '0.4em',
+};
+
+export default function LowerThird(props: LowerProps) {
```

Comment:
the previous solution leveraged a memo to make sure that the component would not re-render unless the relevant props changed (previously presenter and title)

Can we make sure that this is still the case?
I think we would need:
- in automatic mode, check that the lower3 mounts when we go through the events with the << and >> playback
- when a event is running, change the data associated to the event, does the lower3 remount?

Reply:
it looks to work correctly on the screen, if the event id changes the timeout is reset and the data is changed in place
should thees be useMemo or useEffect?

````typescript
const trigger = useMemo(() => {
    if (options.trigger === TriggerType.Event) {
      return eventNow?.id;
    } else if (options.trigger === TriggerType.Manual) {
      return lower.visible;
    }
    return false;
  }, [eventNow?.id, lower.visible, options.trigger]);

  useEffect(() => {
    if (options.trigger === TriggerType.Event && trigger) {
      setPlayState('in');
      const animateOutInMs = options.delay * 1000 + options.transition * 1000;
      const timeout = setTimeout(() => {
        setPlayState('out');
      }, animateOutInMs);
      return () => clearTimeout(timeout);
    } else if (options.trigger === TriggerType.Manual) {
      setPlayState(trigger ? 'in' : 'out');
    } else {
      setPlayState('pre');
    }
    return () => null;
  }, [options.delay, options.transition, options.trigger, trigger]);

`````
![Recording 2024-01-07 013808](https://github.com/cpvalente/ontime/assets/19875125/8a4cf9cb-379c-43d8-938b-43e5c87fa0d1)

Reply:
> should thees be useMemo or useEffect?

`useMemo` and `useEffect` achieve different things.
In this case you can read `useMemo` as in "dont recalculate this object unless one of the dependencies changes value"

where the `useEffect` is more like "run this code when one of the dependencies changes value"

We often avoid using `useEffect`, but in this case we need it to coordinate with the animation. The `useMemo` stabilises the trigger so that it only recalculates if the input data has changed.

As you described, if the id changes, the `useMemo` is triggered, and a new object is created. This, in turn, makes the `useEffect` run, and the animation starts.

With these changes we lose the previous behaviour where the animation was triggered if the data of the lower third changed (title, subtitle, presenter...) But maybe that is good? You have more experience in this field and I will leave it up to you.

We just need to make sure that we dont have issues with race conditions. ie: what happens if eventId changes while we are animating the lower? could we end up with multiple animation transitions scheduled?

Hope this helps